  <div class="alert alert-info">
  Привет, Алексей! Меня зовут Светлана и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

  <div class="alert alert-success">
  <b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
  </div>
  <div class="alert alert-warning">
  <b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
  </div>
  <div class="alert alert-danger">
  <b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
  </div>
  <div class="alert alert-info">
  <b>👂 Совет:</b> Какие-то дополнительные материалы
  </div>
  Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
  Мне будет легче отследить изменения, если ты выделишь свои комментарии:
  <div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
  Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
   </div>

# Анализ книжного сервиса: данные, рейтинги и пользовательские обзоры

**Описание проекта**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цели исследования**

Целью данного исследования является анализ базы данных книжного сервиса, приобретённого компанией, с целью формирования ценностного предложения для нового продукта.

**Ход исследования**

В рамках исследования необходимо:

1. Определить количество книг, опубликованных после 1 января 2000 года.
2. Рассчитать количество обзоров и среднюю оценку для каждой книги.
3. Выявить издательство, выпустившее наибольшее число книг (с учётом только книг толще 50 страниц).
4. Найти автора с самой высокой средней оценкой (учитывая только книги с 50 и более оценками).
5. Определить среднее количество обзоров от пользователей, которые поставили более 48 оценок.

Исследование позволит выявить ключевые тенденции в данных и поможет понять, какие книги наиболее популярны среди пользователей, какие издательства и авторы имеют высокий рейтинг, а также активность пользователей на платформе.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

## Подготовка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [2]:
# Функция для выполнения SQL-запроса и вывода первых строк
def run_query(query):
    return pd.io.sql.read_sql(sql=text(query), con=con)

# Выведем первые 5 строк из каждой таблицы
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

for table in tables:
    print(f"\nТаблица {table}:")
    display(run_query(f"SELECT * FROM {table} LIMIT 5"))


Таблица books:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


Таблица authors:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


Таблица publishers:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


Таблица ratings:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


Таблица reviews:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

 <div class="alert alert-success">
 <b>👍 Успех:</b> Есть описание задачи, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
 </div>

## Выполнение заданий

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [3]:
query = '''
SELECT COUNT(*) AS book_count
  FROM books
 WHERE publication_date > '2000-01-01';
'''

# Выполняем запрос и выводим результат
result = pd.io.sql.read_sql(sql=text(query), con=con)
display(result)

book_count
0         819

**Вывод:**<br/>
После 1 января 2000 года было выпущено **819 книг**.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [4]:
query = '''
SELECT b.title, 
       COUNT(DISTINCT r.review_id) AS review_count, 
       ROUND(AVG(rt.rating), 2) AS avg_rating
  FROM books as b
       LEFT JOIN reviews AS r ON b.book_id = r.book_id
       LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
 GROUP BY b.book_id, b.title
 ORDER BY review_count DESC;
'''

# Выполняем запрос и выводим результат
result = pd.io.sql.read_sql(sql=text(query), con=con)
display(result)

title  review_count  \
0                              Twilight (Twilight  #1)             7   
1                                  Water for Elephants             6   
2                                     The Glass Castle             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...             6   
4    The Curious Incident of the Dog in the Night-Time             6   
..                                                 ...           ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996                            The Natural Way to Draw             0   
997   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998                          Essential Tales and Poems             0   
999  Disney's Beauty and the Beast (A Little Golden...             0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 3 columns]

 <div class="alert alert-danger">
<s> <b>😔 Необходимо исправить:</b> Ты группируешь данные по неуникальному признаку и книг получается меньше, чем в БД. А обзоров получилось слишком много (сколько всего обзоров в БД?), считать нужно только уникальные</s>
 </div>

**Вывод:**<br/>
- Больше всего обзоров у **Twilight (Twilight #1) — 7** при средней оценке **3.66**.

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

In [5]:
query = '''
SELECT 
       p.publisher, 
       COUNT(b.book_id) AS book_count
  FROM books AS b
       JOIN publishers AS p ON b.publisher_id = p.publisher_id
 WHERE b.num_pages > 50
 GROUP BY p.publisher
 ORDER BY book_count DESC
 LIMIT 1;;
'''

# Выполняем запрос и выводим результат
result = pd.io.sql.read_sql(sql=text(query), con=con)
display(result)

publisher  book_count
0  Penguin Books          42

**Вывод:**<br/>
- Наибольшее число книг (**42**) толще 50 страниц выпустило **издательство Penguin Books**.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [6]:
query = '''
SELECT 
       a.author, 
       ROUND(AVG(rating_avg), 2) AS avg_author_rating
  FROM (
        SELECT
               b.author_id, 
               b.title,
               COUNT(rt.rating) AS rating_count, 
               AVG(rt.rating) AS rating_avg
          FROM books AS b
               JOIN ratings AS rt ON b.book_id = rt.book_id
         GROUP BY b.book_id
        HAVING COUNT(rt.rating) >= 50
) AS filtered_books

       JOIN authors AS a ON filtered_books.author_id = a.author_id
 GROUP BY a.author
 ORDER BY avg_author_rating DESC
 LIMIT 1;
'''

# Выполняем запрос и выводим результат
result = pd.io.sql.read_sql(sql=text(query), con=con)
display(result)

author  avg_author_rating
0  J.K. Rowling/Mary GrandPré               4.28

**Вывод:**<br/>
- Автор с самой высокой средней оценкой книг— **J.K. Rowling/Mary GrandPré**.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [7]:
query = '''
WITH active_users AS (
     SELECT username
       FROM ratings
      GROUP BY username
     HAVING COUNT(rating_id) > 48
)

SELECT ROUND(AVG(review_count), 2) AS avg_reviews_per_user
  FROM (
      SELECT r.username, COUNT(r.review_id) AS review_count
        FROM reviews AS r
             JOIN active_users AS au ON r.username = au.username
       GROUP BY r.username
) AS user_reviews;
'''

# Выполняем запрос и выводим результат
result = pd.io.sql.read_sql(sql=text(query), con=con)
display(result)

avg_reviews_per_user
0                  24.0

**Вывод:**<br/>
- Среднее количество обзоров, поставленное пользователями, которые поставили больше 48 оценок — **24**.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

## Общий вывод

1. **Количество книг, выпущенных после 1 января 2000 года:**
Мы определили, что после указанной даты было выпущено 819 книг.

2. **Количество обзоров и средняя оценка книг:**
Изучая обзоры и рейтинги книг, мы обнаружили, что Twilight (Twilight #1) имеет наибольшее количество обзоров — 7, при средней оценке 3.66.

3. **Издательство, выпустившее наибольшее количество книг толще 50 страниц:**
Penguin Books стало лидером по количеству изданных книг (42 книги) объемом более 50 страниц.

4. **Автор с самой высокой средней оценкой книг:**
J.K. Rowling/Mary GrandPré оказалась автором с самой высокой средней оценкой для своих книг.

5. **Среднее количество обзоров от пользователей, поставивших более 48 оценок:**
Мы подсчитали, что пользователи, которые активно оценивают книги (поставили более 48 оценок), оставляют в среднем 24 обзора.

<div class="alert alert-success">
<b>👍 Успех:</b> Хорошая работа, поздравляю с окончанием курса! Рекомендую еще пройти бесплатный курс от практикума по <a href='https://practicum.yandex.ru/ycloud-datalens/'> DataLens</a>
   Дополнительную информацию по SQL  можно посмотреть и потренироваться <a href='https://sql-ex.ru'> здесь</a>, <a href='https://pgexercises.com'> здесь</a>, <a href='https://sql-academy.org'> здесь</a>, <a href='https://stepik.org/course/63054/info'> здесь</a>, <a href='https://Itresume.ru'> здесь</a>
</div>